ttnn/cpp/ttnn/operations/core.hpp

```diff
@@ -168,60 +168,84 @@ inline auto get_memory_config(const ttnn::Tensor& tensor) -> std::optional<ttnn:
                                                : std::optional<ttnn::MemoryConfig>{std::nullopt};
 }
 
+class TensorMonad {
+    const std::vector<Tensor> tensors;
+    const std::vector<std::optional<const Tensor>> optional_tensors;
+
+   public:
+    TensorMonad(
+        const std::vector<Tensor> tensors, const std::vector<std::optional<const Tensor>> optional_tensors = {}) :
+        tensors(std::move(tensors)), optional_tensors(std::move(optional_tensors)) {}
+
+    template <class ConcreteOperation, class... Args>
+    TensorMonad bind(Args... args) {
+        std::vector<Tensor> output_tensors = {
+            Tensor(operation::get_workers_for_op_output(std::move(tensors), std::move(optional_tensors)))};
+        operation::launch_op(
+            [args...](
+                const std::vector<Tensor>& input_tensors,
+                const std::vector<std::optional<const Tensor>>& optional_input_tensors) -> std::vector<Tensor> {
+                return operation::run(
+                    ConcreteOperation(input_tensors, optional_input_tensors, args...),
+                    input_tensors,
+                    optional_input_tensors);
+            },
+            tensors,
+            output_tensors,
+            optional_tensors);
+        return TensorMonad(output_tensors);
+    }
+
+    std::vector<Tensor> get_tensors() { return tensors; }
```

Comment:
do we want to make copies here?

Reply:
We should suggest to the compiler that it should be moved.

Reply:
sounds good